### PR TITLE
Edit setup py to install dev version of required package

### PR DESCRIPTION
**Pull request by praveenkuttappan**
If minimum version of required package as per specifier is not available on PyPI, depends test fails now.

Fix is to identify such requirements and replace specifier to refer dev build version.

### CODE REVIEW

File: eng/tox/sanitize_setuppy.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    for i in range(0, len(contents) - 1):
+        keys = [k for k in requires_dict.keys() if k in contents[i]]
+        for key in keys:
+            contents[i] = contents[i].replace(key, requires_dict[key])
+
+    with open(setup_py_path, "w") as setup_file:
+        setup_file.writelines(contents)
+
+
+def is_required_version_on_pypi(package_name, spec):
```
Comment (scbedd) [inline eng/tox/sanitize_setuppy.py]: The way you use this, you can just `return versions`. A non empty array is `truthy` and an empty array is `falsy`

File: eng/tox/sanitize_setuppy.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        logging.error("setyp.py is not found for package {} to identify current version".format(pkg_name))
+        exit(1)
+
+def process_requires(setup_py_path):
+    # This method process package requirement to verify if all required packages are available on PyPI
+    # If any azure sdk package is not available on PyPI then requirement will be updated to refer dev version
+    requires = [
+        Requirement.parse(r)
+        for r in get_install_requires(setup_py_path)
+        if r.startswith("azure") and "-nspkg" not in r
```
Comment (scbedd) [inline eng/tox/sanitize_setuppy.py]: I'm seeing this a lot. We may want to throw is as a lambda function surfaced in the prospective `azure_devtools.ci_tooling` package we're thinking about.

File: eng/tox/sanitize_setuppy.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    ]
+    # Find package requirements that are not available on PyPI
+    requirement_to_update = {}
+    for req in requires:
+        pkg_name = req.key
+        spec = SpecifierSet(str(req).replace(pkg_name, ""))
+        if not is_required_version_on_pypi(pkg_name, spec):
+            old_req = str(req)
+            requirement_to_update[old_req] = "{0}>={1}".format(
```
Comment (scbedd) [inline eng/tox/sanitize_setuppy.py]: What does this actually do in process? Given an input string of `"azure-core<2.0.0,>=1.2.1";`

It _looks_ to me that it'll become `>= 1.2.1.dev` and blow away the `< 2.0.0`?

File: eng/tox/sanitize_setuppy.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                pkg_name, get_dev_version(pkg_name)
+            )
+    if len(requirement_to_update) == 0:
+        print("All required packages are available on PyPI")
```
Comment (scbedd) [inline eng/tox/sanitize_setuppy.py]: Nit: `print` instead of `logging`

File: eng/tox/sanitize_setuppy.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    requirement_to_update = {}
+    for req in requires:
+        pkg_name = req.key
+        spec = SpecifierSet(str(req).replace(pkg_name, ""))
+        if not is_required_version_on_pypi(pkg_name, spec):
+            old_req = str(req)
+            requirement_to_update[old_req] = "{0}>={1}".format(
+                pkg_name, get_dev_version(pkg_name)
+            )
+    if len(requirement_to_update) == 0:
```
Comment (scbedd) [inline eng/tox/sanitize_setuppy.py]: can just `if requirement_to_update`. 0 length, empty string, None, and any others will just resolve to `true`.

File: eng/tox/sanitize_setuppy.py (hunk 3 above)
Comment (praveenkuttappan) [reply eng/tox/sanitize_setuppy.py]: that's correct. I was worried about any condition that is given like>=1.2.2, <2.0.0 to blindly add .dev. Let me see if i can find the upper bound from specifierset and add

**Pull request merged**